### PR TITLE
feat(PROC-1576): redact pixel data via bounding boxes in edit mode

### PR DESCRIPTION
**Pull request by CalNightingale**
Stacked PR. Depends on #126 (BoundingBox dataclass) and #128 (conftest lift). Comprehensive test coverage in #129.

Lets reviewers permanently black out PHI regions that our de-identification pipeline misses by handing bounding boxes (in raw PixelData coordinates) to COD's existing mode='e'. Unlocks the human-in-the-loop redaction step the review service needs.

Usage mirrors `append`/`edit`: open the series in `mode='e'` and call the method.

```python
with CODObject(client=..., datastore_path=..., study_uid=..., series_uid=..., mode="e") as cod:
    cod.redact_pixel_data(boxes, reviewer="reviewer@gradienthealth.io")
```

- New `CODObject.redact_pixel_data` method (delegates to `redact.py:redact_pixel_data`, same pattern as `append`). Decodes, fills, re-encodes as **JPEG 2000 Lossless** (regardless of source TS), and stamps `BurnedInAnnotation = "NO"` plus a `DeidentificationMethodCodeSequence` entry (DCM 113101 "Clean Pixel Data Option").
- Audit trail: each call appends `{timestamp, reviewer, entries: [...boxes]}` to series-level `metadata_fields["redactions"]`, which survives edit-mode rebuild.
- Hard-fails before any disk write on: missing UID, out-of-range frame, oversized box, or `fill_value` incompatible with `PhotometricInterpretation`/`SamplesPerPixel`. Lock release on raise is unchanged from existing edit-mode behavior, so partial state cannot reach the datastore.
- New `RedactionError` hierarchy in `errors.py`.

**Test scope:** smoke-level only (single-frame happy path, output TS + SOP UID regression, read-mode rejection). Edge-case coverage for audit accumulation, multiframe frame selection, and every hard-error path lives in #129.

**Output TS policy:** redacted instances are normalized to JPEG 2000 Lossless rather than preserving the source transfer syntax. This sidesteps two pydicom behaviors that would otherwise corrupt results: (1) lossy `compress()` regenerates `SOPInstanceUID`, tripping mode='e' set-changed validation; (2) `pixel_array` auto-converts the YBR_FULL family to RGB on decode, so the source PhotometricInterpretation no longer matches the array bytes. Output normalization also matches what `append.py` already does on ingest, so in practice nothing's changing format for data already in COD, and `pylibjpeg-openjpeg` is a hard dep so the encoder is always available.

### CODE REVIEW

**Comment by CalNightingale** [comment]
This change is part of the following stack:

- #126
    - #128
        - #125 ◀
            - #129

<sub>Change managed by [git-spice](https://abhinav.github.io/git-spice/).</sub>

File: cloud_optimized_dicom/redact.py [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+    return (0, 0, 0)
+
+
+def _apply_to_instance(
+    instance: "Instance",
+    boxes: list[BoundingBox],
+    fill_value: Optional[FillValue],
+    now: datetime.datetime,
+) -> None:
+    ds = pydicom3.dcmread(instance.dicom_uri)
```
Comment (saicheems) [inline cloud_optimized_dicom/redact.py]: Looks like we double up on the metadata read, any way to consolidate so we only do the `dcmread` pass once per instance?

File: cloud_optimized_dicom/redact.py [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+    ):
+        framed = arr[np.newaxis, ...]
+    else:
+        framed = arr
+
+    fv = (
+        fill_value
+        if fill_value is not None
+        else _derive_fill_value(samples, effective_pi, bits_stored)
+    )
```
Comment (saicheems) [inline cloud_optimized_dicom/redact.py]:
```suggestion
    fv = fill_value
    if not fv:
        fv = _derive_fill_value(samples, effective_pi, bits_stored)
```

IMO more clear

File: cloud_optimized_dicom/redact.py [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+    seq = list(getattr(ds, "DeidentificationMethodCodeSequence", []) or [])
+    seq.append(method_item)
+    ds.DeidentificationMethodCodeSequence = seq
+
+    # (0008,0012)/(0008,0013) InstanceCreationDate/Time: stamp the instance
+    # with the redaction time. Visible in any DICOM viewer (so a reviewer can
+    # see "this was modified") and used by downstream caches/dedupers that
+    # key on freshness.
+    ds.InstanceCreationDate = now.strftime("%Y%m%d")
+    ds.InstanceCreationTime = now.strftime("%H%M%S")
```
Comment (saicheems) [inline cloud_optimized_dicom/redact.py]: Does this clobber the original timestamps? If so does that matter?

**Review by saicheems — APPROVED** [review]

File: cloud_optimized_dicom/redact.py (hunk 1 above)
Comment (CalNightingale) [reply cloud_optimized_dicom/redact.py]: Done. Pre-flight now does a single `dcmread` per instance and stashes the Dataset in a uid→ds dict for the apply pass to reuse — one open per instance instead of two.

Tried `defer_size=1024` first to keep PixelData lazy, but it interacts badly with `save_as` writing back to the same file: save_as truncates before iterating tags to write, so any deferred tag pydicom hadn't already materialized (e.g., a private tag larger than the threshold) hits "file modification time has changed" on the read-back. Eager read is the simpler safe path; #129's ybr_full_422 fixture happens to reproduce the defer issue, which is how I caught it.

File: cloud_optimized_dicom/redact.py (hunk 2 above)
Comment (CalNightingale) [reply cloud_optimized_dicom/redact.py]: Took the spirit but kept `is None` rather than truthiness. Under `if not fv`, an explicit `fill_value=0` (a valid choice for MONOCHROME1 white-on-black or any caller deliberately picking 0) would silently get replaced by the derived default. Final form is the same shape your suggestion has:

```python
if fill_value is None:
    fill_value = _derive_fill_value(samples, effective_pi, bits_stored)
```

File: cloud_optimized_dicom/redact.py (hunk 3 above)
Comment (CalNightingale) [reply cloud_optimized_dicom/redact.py]: Yes and yes: those describe acquisition/reconstruction time, and the redaction timestamp already lives in the audit record on `metadata_fields["redactions"]`. Dropped the writes; test in #129 now asserts the tags survive the redact unchanged.
